pallets/offchain-rollup/src/anchor.rs

```diff
@@ -0,0 +1,289 @@
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		type OnResponse: OnResponse<Self::AccountId>;
+	}
+
+	/// Anchor response handler trait
+	pub trait OnResponse<AccountId> {
+		fn on_response(name: H256, submitter: AccountId, data: Vec<u8>) -> DispatchResult;
+	}
+	// Default implementation
+	impl<AccountId> OnResponse<AccountId> for () {
+		fn on_response(_name: H256, _submitter: AccountId, _data: Vec<u8>) -> DispatchResult {
+			Ok(())
+		}
+	}
+
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	/// Many-to-one mapping between names and their submitters
+	#[pallet::storage]
+	#[pallet::getter(fn submitter_by_names)]
+	pub type SubmitterByNames<T: Config> = StorageMap<_, Blake2_128Concat, H256, T::AccountId>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn states)]
+	pub type States<T> =
+		StorageDoubleMap<_, Blake2_128Concat, H256, Blake2_128Concat, KeyBytes, ValueBytes>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A name is claimed by a submitter
+		NameClaimed { submitter: T::AccountId, name: H256 },
+		/// A rollup transaction is executed
+		RollupExecuted {
+			submitter: T::AccountId,
+			name: H256,
+			nonce: u128,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// A name was already claimed. You must switch to another name.
+		NameAlreadyClaimed,
+		/// The name doesn't exist
+		NameNotExist,
+		/// The operation is forbidden because it's not done by the name owner
+		NotOwner,
+		/// Rollup condition doesn't meet
+		CondNotMet,
+		/// Cannot decode the action
+		FailedToDecodeAction,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Claims a name and assign the caller as the owner of the name
+		///
+		/// Once the name is claimed, we don't allow to change the owner or deregister any more.
+		#[pallet::weight(0)]
+		#[transactional]
+		pub fn claim_name(origin: OriginFor<T>, name: H256) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			ensure!(
+				SubmitterByNames::<T>::get(&name).is_none(),
+				Error::<T>::NameAlreadyClaimed
+			);
+			SubmitterByNames::<T>::insert(&name, &who);
+			Self::deposit_event(Event::NameClaimed {
+				submitter: who,
+				name,
+			});
+			Ok(())
+		}
+
+		/// Triggers a rollup with an optional nonce
+		#[pallet::weight(0)]
+		#[transactional]
+		pub fn rollup(
+			origin: OriginFor<T>,
+			name: H256,
+			tx: RollupTx,
+			nonce: u128,
+		) -> DispatchResult {
+			// Check submitter
+			let who = ensure_signed(origin)?;
+			Self::ensure_name_owner(&name, &who)?;
+			// Check conditions
+			for cond in tx.conds {
+				let Cond::Eq(key, opt_value) = cond;
+				ensure!(
+					States::<T>::get(name, key) == opt_value,
+					Error::<T>::CondNotMet
+				);
+			}
+			// Exec actions
```

Comment:
Move it down to after the `applying updates`

Reply:
Nice feedback